### ISSUE TITLE
sysdig on Proxmox - sysdig-probe fails to load

### ISSUE THREAD

**Issue by rmoff** [draios/sysdig#415]
I'm trying to run sysdig on Proxmox, and the sysdig-probe fails to load. 
Diagnostics: 

```bash
root@proxmox04:/tmp# pveversion -v
proxmox-ve-2.6.32: 3.4-156 (running kernel: 2.6.32-39-pve)
pve-manager: 3.4-9 (running version: 3.4-9/4b51d87a)
pve-kernel-2.6.32-39-pve: 2.6.32-157
lvm2: 2.02.98-pve4
clvm: 2.02.98-pve4
corosync-pve: 1.4.7-1
openais-pve: 1.1.4-3
libqb0: 0.11.1-2
redhat-cluster-pve: 3.2.0-2
resource-agents-pve: 3.9.2-4
fence-agents-pve: 4.0.10-3
pve-cluster: 3.0-18
qemu-server: 3.4-6
pve-firmware: 1.1-4
libpve-common-perl: 3.0-24
libpve-access-control: 3.0-16
libpve-storage-perl: 3.0-33
pve-libspice-server1: 0.12.4-3
vncterm: 1.1-8
vzctl: 4.0-1pve6
vzprocps: 2.0.11-2
vzquota: 3.1-2
pve-qemu-kvm: 2.2-11
ksm-control-daemon: 1.1-1
glusterfs-client: 3.5.2-1
```

```bash
root@proxmox04:/tmp# uname -r
2.6.32-39-pve
```

```bash
root@proxmox04:/tmp# ls -l /lib/modules/$(uname -r)
total 2368
lrwxrwxrwx  1 root root     36 Aug 14 09:03 build -> /usr/src/linux-headers-2.6.32-39-pve
drwxr-xr-x 12 root root   4096 Jun  5 11:28 kernel
-rw-r--r--  1 root root 605998 May  8 10:40 modules.alias
-rw-r--r--  1 root root 580612 May  8 10:40 modules.alias.bin
-rw-r--r--  1 root root   5665 May  8 10:40 modules.builtin
-rw-r--r--  1 root root   7318 May  8 10:40 modules.builtin.bin
-rw-r--r--  1 root root 205635 May  8 10:40 modules.dep
-rw-r--r--  1 root root 298611 May  8 10:40 modules.dep.bin
-rw-r--r--  1 root root     80 May  8 10:40 modules.devname
-rw-r--r--  1 root root  77081 May  8 10:40 modules.order
-rw-r--r--  1 root root    131 May  8 10:40 modules.softdep
-rw-r--r--  1 root root 273187 May  8 10:40 modules.symbols
-rw-r--r--  1 root root 342206 May  8 10:40 modules.symbols.bin
drwxr-xr-x  2 root root   4096 Jun  5 11:28 zfs
```

```bash
root@proxmox04:/tmp# sysdig-probe-loader
* Unloading sysdig-probe, if present
Error: Module sysdig_probe is not currently loaded
* Running dkms autoinstall

Kernel preparation unnecessary for this kernel.  Skipping...

Building module:
cleaning build area....
make KERNELRELEASE=2.6.32-39-pve -C /lib/modules/2.6.32-39-pve/build M=/var/lib/dkms/sysdig/0.1.102/build....(bad exit status: 2)
Error! Bad return status for module build on kernel: 2.6.32-39-pve (x86_64)
Consult /var/lib/dkms/sysdig/0.1.102/build/make.log for more information.
* Trying to load a system sysdig-probe, if present
* Trying to load a dkms sysdig-probe, if present
* Trying to find precompiled sysdig-probe for 2.6.32-39-pve
Found kernel config at /boot/config-2.6.32-39-pve
* Trying to download precompiled module from https://s3.amazonaws.com/download.draios.com/stable/sysdig-probe-binaries/sysdig-probe-0.1.102-x86_64-2.6.32-39-pve-f9c010e053612af9f0cf3856537bef42.ko
Download failed, consider compiling your own sysdig-probe and loading it or getting in touch with the sysdig community
root@proxmox04:/tmp#
```

```bash
root@proxmox04:/tmp# cat /var/lib/dkms/sysdig/0.1.102/build/make.log
DKMS make.log for sysdig-0.1.102 for kernel 2.6.32-39-pve (x86_64)
Fri Aug 14 09:10:34 BST 2015
expr: syntax error
make: Entering directory `/usr/src/linux-headers-2.6.32-39-pve'
  LD      /var/lib/dkms/sysdig/0.1.102/build/built-in.o
  CC [M]  /var/lib/dkms/sysdig/0.1.102/build/main.o
/var/lib/dkms/sysdig/0.1.102/build/main.c: In function ‘ppm_ioctl’:
/var/lib/dkms/sysdig/0.1.102/build/main.c:782:3: error: implicit declaration of function ‘for_each_process’ [-Werror=implicit-function-declaration]
/var/lib/dkms/sysdig/0.1.102/build/main.c:782:23: error: expected ‘;’ before ‘{’ token
/var/lib/dkms/sysdig/0.1.102/build/main.c:751:27: warning: unused variable ‘t’ [-Wunused-variable]
cc1: some warnings being treated as errors
make[1]: *** [/var/lib/dkms/sysdig/0.1.102/build/main.o] Error 1
make: *** [_module_/var/lib/dkms/sysdig/0.1.102/build] Error 2
make: Leaving directory `/usr/src/linux-headers-2.6.32-39-pve'
root@proxmox04:/tmp#
```

I installed the kernel headers using `apt-get install pve-headers-2.6.32-39-pve` after `apt-get -y install linux-headers-$(uname -r)` failed (`Unable to locate package linux-headers-2.6.32-39-pve`)

What can I do to get this to work? Thanks!

**Comment by ton31337** [comment]
Looks like you don't have debug symbols:
```
implicit declaration of function ‘for_each_process’
```
What does `aptitute search linux-headers-*` return?

**Comment by rmoff** [comment]
```bash
root@proxmox04:/tmp# aptitude search linux-headers-*
v   linux-headers                                                                      -
v   linux-headers-2.6                                                                  -
p   linux-headers-2.6-amd64                                                            - Header files for Linux amd64 configuration (dummy package)
p   linux-headers-3.2.0-4-all                                                          - All header files for Linux 3.2 (meta-package)
p   linux-headers-3.2.0-4-all-amd64                                                    - All header files for Linux 3.2 (meta-package)
i A linux-headers-3.2.0-4-amd64                                                        - Header files for Linux 3.2.0-4-amd64
i A linux-headers-3.2.0-4-common                                                       - Common header files for Linux 3.2.0-4
p   linux-headers-3.2.0-4-common-rt                                                    - Common header files for Linux 3.2.0-4-rt
p   linux-headers-3.2.0-4-rt-amd64                                                     - Header files for Linux 3.2.0-4-rt-amd64
i A linux-headers-amd64                                                                - Header files for Linux amd64 configuration (meta-package)
p   linux-headers-rt-amd64                                                             - Header files for Linux rt-amd64 configuration (meta-package)
```

**Comment by ton31337** [comment]
Try to install `aptitude install linux-headers linux-headers-2.6`

**Comment by rmoff** [comment]
No joy. After doing this: 

```
root@proxmox04:~# aptitude install linux-headers linux-headers-2.6
No packages will be installed, upgraded, or removed.
0 packages upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B of archives. After unpacking 0 B will be used.
```

I still get the same make/build error as above.

**Comment by ton31337** [comment]
strange, did you do `aptitude update` before?

**Comment by rmoff** [comment]
Yep. 
I also tried installing `pve-headers-2.6.32-39-pve` but that didn't help. 

```bash
root@proxmox04:~# aptitude search pve-headers*
p   pve-headers-2.6.32-23-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-24-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-25-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-26-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-27-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-28-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-29-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-30-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-31-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-32-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-33-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-34-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-37-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-2.6.32-38-pve                          - The Proxmox PVE Kernel Headers
i   pve-headers-2.6.32-39-pve                          - The Proxmox PVE Kernel Headers
i   pve-headers-2.6.32-40-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-1-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-10-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-11-pve                          - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-2-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-3-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-4-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-5-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-7-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-8-pve                           - The Proxmox PVE Kernel Headers
p   pve-headers-3.10.0-9-pve                           - The Proxmox PVE Kernel Headers
```

NB I updated the kernel, in case that was part of the problem. 

```bash
root@proxmox04:~# uname -r
2.6.32-40-pve
```

**Comment by gianlucaborello** [comment]
The problem is not related to the missing kernel headers. If they were missing the compilation wouldn't even start. The problem is that your version of the kernel is patched in a way that doesn't provide the `for_each_process` function. We'll take a look but it might take a while since this is probably low priority as it is definitely not a mainstream distribution.

Thanks

**Comment by rmoff** [comment]
Thanks @gianlucaborello.

**Comment by gnosek** [comment]
I hit the same issue on CloudLinux 6.7 (basically patched CentOS with probably-OpenVZ and some other bells and whistles). AIUI, CentOS 6 is supported, so I'd expect to have the for_each_process macro on CloudLinux too. Oh well.

**Comment by jfchevrette** [comment]
It looks like the Proxmox kernel is based off OpenVZ's kernel, which is also the case for Cloudlinux kernel.

The devs at Cloudlinux suggested the following to workaround as it appears the macros used by sysdig have been renamed in OpenVZ's kernel.

    cd /var/lib/dkms/sysdig/0.6.0/source/
    sed -i 's/for_each_process/for_each_process_all/g' *.[ch]
    sed -i 's/while_each_thread/while_each_thread_all/g' *.[ch]
    sed -i 's/do_each_thread/do_each_thread_all/g' *.[ch]

Then run the following to build and install the module
    dkms build -m sysdig -v 0.6.0
    dkms install --force -m sysdig -v 0.6.0

Worked for me and I can start sysdig/csysdig and trace stuff. YMMV.

Any chances the sysdig folks can see if and how this could be handled in sysdig's source?

**Comment by Mikepoplas** [comment]
Hi, I've tried to apply this to version 0.13.0 but:

DKMS make.log for sysdig-0.13.0 for kernel 4.15.18-13-pve (x86_64)
Fri May  3 10:54:13 CEST 2019
make: Entering directory '/usr/src/linux-headers-4.15.18-13-pve'
Makefile:976: "Cannot use CONFIG_STACK_VALIDATION=y, please install libelf-dev, libelf-devel or elfutils-libelf-devel"
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/main.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/dynamic_params_table.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/flags_table.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/ppm_events.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/ppm_fillers.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/event_table.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/syscall_table.o
  CC [M]  /var/lib/dkms/sysdig/0.13.0/build/ppm_cputime.o
/var/lib/dkms/sysdig/0.13.0/build/main.c:121:55: error: unknown type name ‘cputime_t’
 void ppm_task_cputime_adjusted(struct task_struct *p, cputime_t *ut, cputime_t *st);
                                                       ^~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c:121:70: error: unknown type name ‘cputime_t’
 void ppm_task_cputime_adjusted(struct task_struct *p, cputime_t *ut, cputime_t *st);
                                                                      ^~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c: In function ‘ppm_ioctl’:
/var/lib/dkms/sysdig/0.13.0/build/main.c:616:3: error: implicit declaration of function ‘for_each_process_all’ [-Werror=implicit-function-declaration]
   for_each_process_all(p) {
   ^~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c:616:27: error: expected ‘;’ before ‘{’ token
   for_each_process_all(p) {
                           ^
/var/lib/dkms/sysdig/0.13.0/build/main.c:587:27: warning: unused variable ‘t’ [-Wunused-variable]
   struct task_struct *p, *t;
                           ^
/var/lib/dkms/sysdig/0.13.0/build/main.c: In function ‘cpu_callback’:
/var/lib/dkms/sysdig/0.13.0/build/main.c:1975:7: error: ‘CPU_UP_PREPARE_FROZEN’ undeclared (first use in this function)
  case CPU_UP_PREPARE_FROZEN:
       ^~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c:1975:7: note: each undeclared identifier is reported only once for each function it appears in
/var/lib/dkms/sysdig/0.13.0/build/main.c:1979:7: error: ‘CPU_DOWN_PREPARE’ undeclared (first use in this function)
  case CPU_DOWN_PREPARE:
       ^~~~~~~~~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c:1981:7: error: ‘CPU_DOWN_PREPARE_FROZEN’ undeclared (first use in this function)
  case CPU_DOWN_PREPARE_FROZEN:
       ^~~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c: In function ‘sysdig_init’:
/var/lib/dkms/sysdig/0.13.0/build/main.c:2132:2: error: implicit declaration of function ‘register_cpu_notifier’ [-Werror=implicit-function-declaration]
  register_cpu_notifier(&cpu_notifier);
  ^~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/sysdig/0.13.0/build/main.c: In function ‘sysdig_exit’:
/var/lib/dkms/sysdig/0.13.0/build/main.c:2181:2: error: implicit declaration of function ‘unregister_cpu_notifier’ [-Werror=implicit-function-declaration]
  unregister_cpu_notifier(&cpu_notifier);
  ^~~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
scripts/Makefile.build:322: recipe for target '/var/lib/dkms/sysdig/0.13.0/build/main.o' failed
make[1]: *** [/var/lib/dkms/sysdig/0.13.0/build/main.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile:1552: recipe for target '_module_/var/lib/dkms/sysdig/0.13.0/build' failed
make: *** [_module_/var/lib/dkms/sysdig/0.13.0/build] Error 2
make: Leaving directory '/usr/src/linux-headers-4.15.18-13-pve'

Is still possible to run in proxmox with current kernel?

Regards
